### PR TITLE
Optimize handle_response cache check

### PR DESCRIPTION
**Pull request by bdraco**
Zeroconf was eating up a bit more cpu time then expected with lots of devices on the network.

Reduce the number of calls to __eq__ when checking the cache as they are far more expensive than the 1000 second check.

Before
Screen Shot 2020-02-29 at 11 14 30 AM

After
Screen Shot 2020-02-29 at 11 15 17 AM

### CODE REVIEW

**Comment by jstasiak** [comment]
Hey, thanks for this, it looks reasonable. A commit message needs to be provided with the description of the test environment, the way this was tested and benchmarks/comparison with all relevant units included (benchmarks in text, not images).

**Comment by bdraco** [comment]
@jstasiak Commit message updated and repushed

**Comment by jstasiak** [comment]
Thanks. Can you provide the actual timings of `handle_response` before and after? It's tricky to judge the improvement by the percentages alone.

**Comment by bdraco** [comment]
@jstasiak Unfortunately py-spy doesn't have a way to report the data as timings, only a percentage of sample time.

**Comment by bdraco** [comment]
That  is a lot better.  After 10200 samples handle_response isn't even in the top few calls

```
Collecting samples from '/usr/local/bin/python3 -m homeassistant --config /config' (python v3.7.6)
Total Samples 10200
GIL: 7.00%, Active: 61.00%, Threads: 35

  %Own   %Total  OwnTime  TotalTime  Function (filename:line)                                                                                                                       
 47.00%  47.00%   24.84s    24.84s   _worker (concurrent/futures/thread.py:78)
  5.00%   5.00%    2.97s     2.97s   run (zeroconf/__init__.py:1226)
  1.00%   1.00%   0.390s    0.390s   handle_read (zeroconf/__init__.py:1265)
  1.00%   1.00%   0.200s    0.200s   read_utf (zeroconf/__init__.py:818)
  0.00%   0.00%   0.120s    0.120s   unpack (zeroconf/__init__.py:723)
  0.00%   1.00%   0.120s    0.320s   read_name (zeroconf/__init__.py:834)
  0.00%   0.00%   0.100s    0.240s   update_record (zeroconf/__init__.py:2440)
  0.00%   0.00%   0.090s    0.090s   notify (threading.py:352)
  0.00%   0.00%   0.070s    0.070s   update_record (zeroconf/__init__.py:1469)
  0.00%   0.00%   0.060s    0.070s   __eq__ (zeroconf/__init__.py:606)
  0.00%   0.00%   0.050s    0.050s   acquire (logging/__init__.py:843)
  0.00%   0.00%   0.050s    0.050s   unpack (zeroconf/__init__.py:722)
  0.00%   0.00%   0.050s    0.050s   read_name (zeroconf/__init__.py:828)
  0.00%   0.00%   0.050s    0.050s   is_expired (zeroconf/__init__.py:494)
  0.00%   0.00%   0.040s    0.040s   emit (logging/__init__.py:1028)
  1.00%   1.00%   0.040s    0.040s   __init__ (zeroconf/__init__.py:386)
  0.00%   0.00%   0.040s    0.040s   __enter__ (threading.py:241)
```

File: zeroconf/__init__.py [hunk 1]
```diff
@@ -2444,13 +2449,22 @@ def handle_response(self, msg: DNSIncoming) -> None:
...
-                    if entry == record:
-                        updated = False
-
-                    if DNSEntry.__eq__(entry, record) and (record.created - entry.created > 1000):
-                        self.cache.remove(entry)
+                # Since the cache format is keyed on DNSEntry
+                # we can avoid iterating everything in the cache and
+                # only look though entries for the specific DNSEntry
+                # cache key.
+                entries = self.cache.entries_with_name(DNSEntry(record.name, record.type, record.class_).key)
```
Comment (jstasiak) [inline zeroconf/__init__.py]: `DNSEntry(record.name, record.type, record.class_).key` just returns `record.name.lower()`, but `entries_with_name()` already performs `lower()` on the name passed, so `DNSEntry(record.name, record.type, record.class_).key` should be just `record.name` I think.

File: zeroconf/__init__.py [hunk 2]
```diff
@@ -2444,13 +2449,22 @@ def handle_response(self, msg: DNSIncoming) -> None:
...
+                if entries:
```
Comment (jstasiak) [inline zeroconf/__init__.py]: This condition is unnecessary, if the list is empty the loop below will just have zero iterations.

File: zeroconf/__init__.py [hunk 3]
```diff
@@ -2444,13 +2449,22 @@ def handle_response(self, msg: DNSIncoming) -> None:
...
+                    # Make a copy since we are likely to delete
+                    # and we cannot iterate over something we
+                    # might remove
+                    for entry in entries.copy():
+
+                        if entry == record:
+                            updated = False
+
+                        if (record.created - entry.created > 1000) and DNSEntry.__eq__(entry, record):
```
Comment (jstasiak) [inline zeroconf/__init__.py]: Is this optimization (reordering of the expressions) still necessary? If so please leave a comment mentioning this and include it in the commit message, otherwise it's not really visible that this changed.

File: zeroconf/__init__.py [hunk 4]
```diff
@@ -1136,6 +1136,11 @@ def remove(self, entry: DNSRecord) -> None:
         try:
             list_ = self.cache[entry.key]
             list_.remove(entry)
+            # If we remove the last entry in the list
+            # we remove the key from the dict in order
+            # to avoid leaking memory
+            if not list_:
+                del self.cache[entry.key]
```
Comment (jstasiak) [inline zeroconf/__init__.py]: While being cool to have this addressed, this belongs in a separate pull request, it's not a `handle_response()` optimization.

File: zeroconf/__init__.py (hunk 1 above)
Comment (bdraco) [reply zeroconf/__init__.py]:
```
2020-03-07 11:44:27 WARNING (zeroconf-Engine) [zeroconf] RECORD CHECK: dnsentry_key=[wiced-hap-3e2734.local.] name=[WICED-hap-3E2734.local.]
2020-03-07 11:44:27 WARNING (zeroconf-Engine) [zeroconf] RECORD CHECK: dnsentry_key=[wiced-hap-3e2734.local.] name=[WICED-hap-3E2734.local.]
```

Confirmed

File: zeroconf/__init__.py (hunk 2 above)
Comment (bdraco) [reply zeroconf/__init__.py]: Confirmed

```
        except KeyError:
            return []
``` 

Returns an empty list on error

File: zeroconf/__init__.py (hunk 3 above)
Comment (bdraco) [reply zeroconf/__init__.py]: The > is still far cheaper than making the __eq__ call so will note that

File: zeroconf/__init__.py (hunk 4 above)
Comment (bdraco) [reply zeroconf/__init__.py]: Will make a new PR for this after this one is sorted out

**Comment by bdraco** [comment]
> Few minor comments, but nice work. Can you include the text benchmarks in the commit message together with the name of the software used to produce them (is it still py-spy?)?

Yes, still py-spy.  I found another flag that gave me text output (added to the PR commit message)

I retested with the changes and handle_response doesn't show up in the profile anymore so no  measurable change to before/after (although I'm sure its a bit faster without creating the DNSEntry and calling .key).

**Comment by jstasiak** [comment]
Nice, cheers!
